### PR TITLE
Retain order when deduplicating message types

### PR DESCRIPTION
**Pull request by timbussmann**
fixes https://github.com/Particular/NServiceBus/issues/4794

Decided to not remove the deduplication code altogether at this point as the current behavior is quite hard to follow for certain edge cases related to complex inheritance chains.

### CODE REVIEW

File: src/NServiceBus.Core/Pipeline/Outgoing/SerializeMessageConnector.cs [hunk 1]
```diff
@@ -52,10 +53,16 @@ string SerializeEnclosedMessageTypes(Type messageType)
         {
             var metadata = messageMetadataRegistry.GetMessageMetadata(messageType);
 
-            var assemblyQualifiedNames = new HashSet<string>();
+            var assemblyQualifiedNames = new List<string>();
             foreach (var type in metadata.MessageHierarchy)
             {
-                assemblyQualifiedNames.Add(type.AssemblyQualifiedName);
+                var typeAssemblyQualifiedName = type.AssemblyQualifiedName;
+                if (assemblyQualifiedNames.Contains(typeAssemblyQualifiedName))
```
Comment (bording) [inline src/NServiceBus.Core/Pipeline/Outgoing/SerializeMessageConnector.cs]: Since we're likely dealing with relatively low numbers of items in `metadata.MessageHierarchy`, I suppose adding an O(N) call here isn't too big of a deal.

File: src/NServiceBus.Core/Pipeline/Outgoing/SerializeMessageConnector.cs [hunk 2]
```diff
@@ -52,10 +53,16 @@ string SerializeEnclosedMessageTypes(Type messageType)
         {
             var metadata = messageMetadataRegistry.GetMessageMetadata(messageType);
 
-            var assemblyQualifiedNames = new HashSet<string>();
+            var assemblyQualifiedNames = new List<string>();
```
Comment (bording) [inline src/NServiceBus.Core/Pipeline/Outgoing/SerializeMessageConnector.cs]: Since we know the upper bound on the number of items, we should initialize the capacity to `metadata.MessageHierarchy.Length` to avoid multiple internal array allocations.

File: src/NServiceBus.Core/Pipeline/Outgoing/SerializeMessageConnector.cs [hunk 3]
```diff
@@ -3,6 +3,7 @@ namespace NServiceBus
     using System;
     using System.Collections.Generic;
     using System.IO;
+    using System.Linq;
```
Comment (bording) [inline src/NServiceBus.Core/Pipeline/Outgoing/SerializeMessageConnector.cs]: :fire:

File: src/NServiceBus.Core/Pipeline/Outgoing/SerializeMessageConnector.cs (hunk 2 above)
Comment (timbussmann) [reply src/NServiceBus.Core/Pipeline/Outgoing/SerializeMessageConnector.cs]: yeah was thinking about this as well. Since I assumed that in most cases the hierarchy is quite small, the default size is a suitable default value but since the amount of duplicate message types should be super low, specifying a size seems to make sense. Will add that.

File: src/NServiceBus.Core/Pipeline/Outgoing/SerializeMessageConnector.cs (hunk 1 above)
Comment (timbussmann) [reply src/NServiceBus.Core/Pipeline/Outgoing/SerializeMessageConnector.cs]: The alternative would be to use `Distinct` which works with an internal set for the deduplication check, but since the expected amount of entries in these collections should be very small (< 5), I think this shouldn't be any issue.

Comment (danielmarbach) [reply src/NServiceBus.Core/Pipeline/Outgoing/SerializeMessageConnector.cs]: For reference the PR which introduced this https://github.com/Particular/NServiceBus/pull/4009

Comment (danielmarbach) [reply src/NServiceBus.Core/Pipeline/Outgoing/SerializeMessageConnector.cs]: How about doing a distinct in https://github.com/Particular/NServiceBus/blob/develop/src/NServiceBus.Core/Unicast/Messages/MessageMetadataRegistry.cs#L113 or would that break other things?
